File: frontend/src/api/payment.ts
import { apiClient } from './client';

/**
 * SavedCard represents a card stored with Square
 *
 * id = card_id from Square API (format: "card_xxx")
 * Never use the nonce (cnon_xxx) for saved cards
 */
export interface SavedCard {
  id: string; // card_id from Square (card_xxx), NOT sourceId (cnon_xxx)
  cardholderName?: string; // ⭐ Cardholder name
  lastFourDigits: string;
  cardType: string; // "VISA", "MASTERCARD", "AMEX"
  expiryMonth: number;
  expiryYear: number;
  isDefault: boolean;
  createdAt?: string;
}

export interface GetSavedCardsResponse {
  success: boolean;
  message: string;
  data?: SavedCard[];
}

export interface AddCardRequest {
  sourceId: string; // Payment nonce from Square Web Payments SDK (cnon_xxx)
  cardholderName: string; // ⭐ Cardholder name for card storage
  squareCustomerId?: string; // ⭐ Square Customer ID (optional if already in DB)
  verificationToken?: string; // ⭐ From verifyBuyer() for SCA
  billingAddress?: {
    givenName?: string;
    familyName?: string;
    addressLine1?: string;
    addressLine2?: string;
    administrativeDistrictLevel1?: string;
    postalCode?: string;
    country?: string;
  };
}

export interface AddCardResponse {
  success: boolean;
  message: string;
  data?: SavedCard;
}

export interface DeleteCardResponse {
  success: boolean;
  message: string;
}

export interface SetDefaultCardResponse {
  success: boolean;
  message: string;
  data?: SavedCard;
}

export interface PaymentRequest {
  sourceId: string;
  amount: number;
  currency?: string;
  orderId?: string;
}

export interface PaymentResponse {
  id: string;
  status: string;
  receipt_number?: string;
  receipt_url?: string;
  [key: string]: any;
}

interface SubmitPaymentApiResponse {
  success: boolean;
  message?: string;
  error?: string;
  data?: PaymentResponse;
}

/**
 * Get user's saved cards
 */
export async function getSavedCards(): Promise<GetSavedCardsResponse> {
  return apiClient.get<GetSavedCardsResponse>('/payment-methods');
}

/**
 * Add new card
 */
export async function addCard(
  cardData: AddCardRequest,
): Promise<AddCardResponse> {
  // ⭐ 診断: cardData の内容をログ出力
  console.log('[DIAGNOSTIC] addCard called with:', cardData);
  console.log('[DIAGNOSTIC] sourceId:', cardData.sourceId);
  console.log('[DIAGNOSTIC] sourceId type:', typeof cardData.sourceId);
  console.log('[DIAGNOSTIC] sourceId length:', cardData.sourceId?.length);
  console.log('[DIAGNOSTIC] Full cardData keys:', Object.keys(cardData));

  // ⭐ 重要: sourceId、cardholderName、verificationToken のみをサーバーに送信
  // brand, last4, expMonth, expYear は一切送らない
  const cleanData: AddCardRequest = {
    sourceId: cardData.sourceId,
    cardholderName: cardData.cardholderName,
    verificationToken: cardData.verificationToken,
    billingAddress: cardData.billingAddress,
  };

  console.log('[DIAGNOSTIC] Cleaned cardData for sending:', cleanData);
  console.log('[DIAGNOSTIC] Cleaned keys:', Object.keys(cleanData));

  return apiClient.post<AddCardResponse>('/payment-methods', cleanData);
}

/**
 * Delete card
 */
export async function deleteCard(cardId: string): Promise<DeleteCardResponse> {
  return apiClient.delete<DeleteCardResponse>(`/payment-methods/${cardId}`);
}

/**
 * Set card as default
 */
export async function setDefaultCard(
  cardId: string,
): Promise<SetDefaultCardResponse> {
  return apiClient.put<SetDefaultCardResponse>(
    `/payment-methods/${cardId}/default`,
    {},
  );
}

/**
 * Submit payment to backend API (client-side)
 */
export async function submitPayment(
  paymentRequest: PaymentRequest,
  authToken?: string,
): Promise<PaymentResponse> {
  const { sourceId, amount, currency = 'JPY', orderId } = paymentRequest;

  const options: RequestInit = authToken
    ? {
        headers: {
          Authorization: `Bearer ${authToken}`,
        },
      }
    : {};

  const response = await apiClient.post<SubmitPaymentApiResponse>(
    '/payments',
    {
      sourceId,
      amount: Math.round(amount),
      currency,
      orderId,
    },
    options,
  );

  if (!response.success || !response.data) {
    throw new Error(
      response.message || response.error || 'Payment processing failed',
    );
  }

  return response.data;
}

File: backend/src/models/payment.py
"""
Payment Method Model for DynamoDB
"""


class PaymentMethod:
    """Payment method entity"""

    def __init__(
        self,
        payment_method_id: str,
        brand: str,
        last4: str,
        exp_month: int,
        exp_year: int,
        is_main: bool = False,
        status: str = "active"
    ):
        self.payment_method_id = payment_method_id
        self.brand = brand
        self.last4 = last4
        self.exp_month = exp_month
        self.exp_year = exp_year
        self.is_main = is_main
        self.status = status

    def to_dict(self):
        """Convert to dictionary for API response"""
        return {
            "id": self.payment_method_id,
            "cardType": self.brand,
            "lastFourDigits": self.last4,
            "expiryMonth": self.exp_month,
            "expiryYear": self.exp_year,
            "isDefault": self.is_main
        }

    @staticmethod
    def from_dynamo(item: dict) -> "PaymentMethod":
        """Create from DynamoDB item"""
        return PaymentMethod(
            payment_method_id=item.get("paymentMethodId"),
            brand=item.get("brand"),
            last4=item.get("last4"),
            exp_month=int(item.get("expMonth", 0)),
            exp_year=int(item.get("expYear", 0)),
            is_main=item.get("isMain", False),
            status=item.get("status", "active")
        )

    @staticmethod
    def to_dynamo_item(user_id: str, payment_method: "PaymentMethod") -> dict:
        """Convert to DynamoDB item format"""
        return {
            "PK": f"USER#{user_id}",
            "SK": f"PAYMENT_METHOD#{payment_method.payment_method_id}",
            "paymentMethodId": payment_method.payment_method_id,
            "brand": payment_method.brand,
            "last4": payment_method.last4,
            "expMonth": payment_method.exp_month,
            "expYear": payment_method.exp_year,
            "isMain": payment_method.is_main,
            "status": payment_method.status
        }
